2019-08-03-q-and-as-about-nix-for-haskellers.md

```diff
@@ -0,0 +1,692 @@
+
+```nix
+stdenv.mkDerivation {
+  name = "hello-2.1.1";
+  builder = ./builder.sh;
+  src = fetchurl {
+    url = ftp://ftp.nluug.nl/pub/gnu/hello/hello-2.1.1.tar.gz;
+    sha256 = "1md7jsfd8pa45z73bz1kszpp01yw6x5ljkjk2hx7wl800any6465";
+  };
+```
+
+As far as the Nix language is concerned, you can think of a derivation as a
+dictionary of key-value pairs.  Some of the keys (and corresponding values) are
+considered "special".  For instance, in the above definition, `builder` is a
+special key that corresponds to a shell script.  The shell script contains
+commands like `./configure && make && make install` that will be used to
+build and install the package.
+
+Nix has a few well-known CLI tools.  In this article I will mostly talk about
+`nix-build` and `nix-shell`.  `nix-build` would take the above derivation and
+actually build it.  `nix-build` uses the "special" keys in the
+above derivation to run shell scripts to build and install the binaries, shared
+objects, header files, and other files associated with the package.[^2]
+
+For instance, if the above derivation is saved in a file called
+`my-first-derivation.nix`, you can build it with the following command:[^3]
+
+```
+$ nix-build ./my-first-derivation.nix
+```
+
+This produces a directory in the Nix store called something like
+`/nix/store/06vykrz1hmxgxir8i74fwjl6r9bb2gpg-hello-2.10`.  This contains, among
+other things, the `hello` binary:
+
+```
+$ /nix/store/06vykrz1hmxgxir8i74fwjl6r9bb2gpg-hello-2.10/bin/hello
+Hello, world!
+```
+
+*That doesn't sound too bad.  Where is the difficulty?*
+
+If this was all there is to Nix, it wouldn't be too difficult.  However, it
+also wouldn't be that helpful.
+
+The other big part of Nix is a huge set of pre-defined derivations and helper
+functions in a repo called [nixpkgs](https://github.com/NixOS/nixpkgs).
+This repository contains tens of thousands of derivations, as well as hundreds
+of helper functions.  In practice, almost no one uses the bare Nix language by
+itself.  Almost everyone uses Nix with nixpkgs.
+
+So what makes learning Nix hard?  In my opinion, it is the following three things:
+
+1.  The documentation for Nix is both terrible and great at the same time.  The
+    main documentation for the Nix language and CLI tools is in the
+    [Nix manual](https://nixos.org/nix/manual/).  The main documentation for the
+    derivations and helper functions in nixpkgs is the
+    [nixpkgs manual](https://nixos.org/nixpkgs/manual/).
+
+    For the most part, these manuals are written as reference manuals rather
+    than beginner-oriented HOW-TO manuals. As a beginner, it is quite difficult
+    to figure out how to do things just from the above manuals.[^4]
+
+2.  Nix is a dynamic language.  Unlike Haskell, you don't get "documentation for
+    free" with type signatures.
+
+    There are no tools similar to [Haddock](https://www.haskell.org/haddock/)
+    or [doctest](https://github.com/sol/doctest) for Nix.
```

Comment:
Minor nitpicking: probably absence of the `doctest` is not that important. `doctest` is for automatically testing examples in the documentation. Probably won't be useful. But the absence of something like Hoogle is much more crucial. How am I supposed to know what a function does or a field means if I don't have Hoogle? :see_no_evil: